### ISSUE TITLE
How to disable buffer title? Possible to add icons to crumbs?

### ISSUE THREAD

**Issue by Icy-Thought** [joaotavora/breadcrumb#27]
Greetings! 

1. Thank you for creating this package and making it publicly available for the rest of us to use!
2. I failed to find a way, with my very limited lisp knowledge, to hide the buffer-title in the `breadcrumb` and therefore I thought I should ask here if it was possible to disable/hide it in any way.
3. Is it possible to add images (`kind-icons`?) to` breadcrumb`? 
    + I have noticed sometimes that `breadcrumb` lacks some depth, don't know the cause of this tbh.

Example code block:

```nix
    overlays =
      (mapModules ./overlays import)
      // {
        default = final: prev: {
          unstable = pkgs-unstable;
          my = self.packages.${system};
        };
      };
```

What I see in neovim:

![Selection 2024-01-02 20:36:33](https://github.com/joaotavora/breadcrumb/assets/53710398/a44891f2-e937-4d80-aaec-a5e733af6e3c)

Versus what I see in Emacs:

![Selection 2024-01-02 20:40:20](https://github.com/joaotavora/breadcrumb/assets/53710398/a85203d7-e7b3-40ff-84b0-914e1dbbaeda)

**Comment by nqminhuit** [comment]
> I failed to find a way, with my very limited lisp knowledge, to hide the buffer-title in the breadcrumb and therefore I thought I should ask here if it was possible to disable/hide it in any way.

for me, I do it this way:
```elisp
(fset 'breadcrumb--project-crumbs-1 #'ignore)
```

**Comment by Icy-Thought** [comment]
I had forgot about this until you just reminded me of the issue. Thank you for sharing the solution with us! :)

**Comment by mrcnski** [comment]
> > I failed to find a way, with my very limited lisp knowledge, to hide the buffer-title in the breadcrumb and therefore I thought I should ask here if it was possible to disable/hide it in any way.
> 
> for me, I do it this way:
> 
> (fset 'breadcrumb--project-crumbs-1 #'ignore)

Thank you! There could be a variable to easily configure this. :)

**Comment by joaotavora** [comment]
Probably easier to read the few words of https://github.com/joaotavora/breadcrumb#more-usage and understand  what's written there

**Comment by kantholtz** [comment]
TLDR:

```elisp
(setq-default header-line-format
              '((:eval (breadcrumb-imenu-crumbs)))))
```

However, doing it this way, an empty header-line is always shown for all buffers. This deviates from the standard behaviour iirc...

## Understanding What's Written There

> Probably easier to read the few words of https://github.com/joaotavora/breadcrumb#more-usage and understand what's written there

For other people with limited elisp knowledge like me, I append this write-up of my attempt at understanding how it works. Everything is straightforward in hindsight, but maybe this helps people out figuring out stuff on their own:

> (...) if you want some leet modeline (...) put the mode-line constructs (:eval (breadcrumb-imenu-crumbs))

As I have not modified my header-line so far, let's see how this is done. The documentation for `header-line-format` says, among others (`C-h v header-line-format RET`):

> Analogous to ‘mode-line-format’, but controls the header line. The header line appears, optionally, at the top of a window; the mode line appears at the bottom.

So, to understand the `(:eval ...)` construct, it seems we need to look up the documentation for `mode-line-format` with `C-h v mode-line-format RET`), and yes, it tells us that the `mode-line` (and therefore also the `header-line`) is

> (...) a variable (...) [whose] value may be nil, a string, a symbol or a list.

We need to assume that the code examples in the "more usage" section of the readme describe elements of the `header-line-format` defined as a list and

> A list whose **car** is a string or list is processed by processing each of the list elements recursively, as separate **mode line constructs**, and concatenating the results.

There is the familiar term **mode line construct** we find in the README. So if we construct the header-line-format as a list, and if its **first element** (car) is a string or a list, then all elements are evaluated and concatenated to produce the final header line. The documentation then also shows that `(:eval FORM)` is a valid list to be used inside the `header-line-format` and `mode-line-format`.

Alright, up to this point, we learned to do this:

```elisp
(setq-default header-line-format
              '((:eval (breadcrumb-project-crumbs))
                (:eval (breadcrumb-imenu-crumbs))))
```

From what I've read, it's `setq-default` and not `setq` since the header-line and mode-line are buffer-local variables for which `setq` would not apply to other buffers apart from itself. Let's see what `breadcrumb-imenu-crumbs` does by `C-h f breadcrumb-imenu-crumbs RET`:

> Describe point inside the Imenu tree of current file.

Ok, so `breadcrumb-mode` uses the (built-in) [IMenu](https://www.gnu.org/software/emacs/manual/html_node/emacs/Imenu.html) to figure out what the crumbs should be. Setting up a test file `test.el` with the following content:

```elisp
(defun test-fn ()
  (breadcrumb-project-crumbs)
  (breadcrumb-imenu-crumbs))
```

and evaluating the return value of `breadcrumb-imenu-crumbs` with `C-x C-e` yields

```#("test-fn" 0 1 (breadcrumb-siblings nil keymap (keymap (mode-line keymap ...) (header-line keymap ...)) help-echo #("mouse-1: Go places near test-fn" 24 25 (breadcrumb-siblings nil)) breadcrumb-dont-shorten t face breadcrumb-imenu-leaf-face mouse-face header-line-highlight) 1 7 (keymap (keymap (mode-line keymap ...) (header-line keymap ...)) help-echo #("mouse-1: Go places near test-fn" 24 25 (breadcrumb-siblings nil)) breadcrumb-dont-shorten t face breadcrumb-imenu-leaf-face mouse-face header-line-highlight)) ```

This is a so-called `propertized string` which is marked by `#(...)`. From what I understand, it contains (1) the actual string to be shown, (2) styling and interaction options applied to character-ranges (i.e. the numbers, for example 0 1, 24 25, etc., as seen above). This specific propertized string appears to contain the information that allows us to click on a crumb to jump to the definition. The two functions yield

1. `breadcrumb-imenu-crumbs`: `#("test-fn" ...)` *The position of the cursor inside the imenu tree*
2. `breadcrump-project-crumbs`: `#("test.el" ...)` *The file or project name*

So if we only use (`breadcrumb-imenu-crumbs`) this should remove the filename at the beginning of the header line.

**Comment by joaotavora** [comment]
> Understanding What's Written There

Wow, that was some deep dive into some arguably very irrelevant implementation details.   But kudos nonetheless.  I guess when learning how to do an  omelette you first investigate the denaturation of albumin and the meaning of life in general :-)

 Anyway,  I've added `M-x breadcrumb-opinionated-mode` , it comes with a related variable `breadcrumb-opinionated-mlf` that you can customize and  see how things are put into and removed from lists. 

> However, doing it this way, an empty header-line is always shown for all buffers. This deviates from the standard behaviour iirc...

There's little I can do about that, since the header line is only elided completely when it is entirely nil which equates in Elisp with an empty list, which by definition you don't want (you want _something_ to be there).
